### PR TITLE
Enhance LDAP authentication

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

This PR proposes to enhance the LDAP support, which mainly referring the code introduced in HIVE-14713.

Currently, Kyuubi has very limited LDAP support, and the implementation is from the early Hive codebase. Hive enhanced the LDAP support in later versions, considering the Hive ecosystem is quite mature, I think we'd better to porting this functionality and keep the same behavior w/ Hive first, and we can improve it if meet certain requirements/issues in the future.

Basically, this PR introduces the following configurations

```
kyuubi.authentication.ldap.url (since 1.0.0)
kyuubi.authentication.ldap.domain (since 1.0.0)
kyuubi.authentication.ldap.guidKey (since 1.2.0)
kyuubi.authentication.ldap.base.dn (since 1.0.0 deprecated)
kyuubi.authentication.ldap.baseDN
kyuubi.authentication.ldap.groupMembershipKey
kyuubi.authentication.ldap.userMembershipKey
kyuubi.authentication.ldap.groupClassKey
kyuubi.authentication.ldap.groupDNPattern
kyuubi.authentication.ldap.userDNPattern
kyuubi.authentication.ldap.groupFilter
kyuubi.authentication.ldap.userFilter
kyuubi.authentication.ldap.customLDAPQuery
kyuubi.authentication.ldap.binddn
kyuubi.authentication.ldap.bindpw
```

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

This PR ports all LDAP-related UT&IT from Hive codebase

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -783,8 +783,9 @@ object KyuubiConf {
       .stringConf
       .createOptional
 
-  val AUTHENTICATION_LDAP_BASEDN: OptionalConfigEntry[String] =
-    buildConf("kyuubi.authentication.ldap.base.dn")
+  val AUTHENTICATION_LDAP_BASE_DN: OptionalConfigEntry[String] =
+    buildConf("kyuubi.authentication.ldap.baseDN")
+      .withAlternative("kyuubi.authentication.ldap.base.dn")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:788]: Keep consistent w/ Hive keys

File: .gitignore [hunk 2]
```diff
@@ -32,11 +32,7 @@
 .ensime_lucene
 .generated-mima*
 .vscode/
-# The star is required for further !/.idea/ to work, see https://git-scm.com/docs/gitignore
-/.idea/*
-# Icon for JetBrains Toolbox
-!/.idea/icon.png
-!/.idea/vcs.xml
```
Comment (pan3793) [inline .gitignore:39]: It's redundant because we can add arbitrary files using `git add --force`

File: .gitignore [hunk 3]
```diff
@@ -59,7 +55,6 @@ hs_err_pid*
 spark-warehouse/
 metastore_db
 derby.log
-ldap
```
Comment (pan3793) [inline .gitignore:62]: I don't see any temporary folder names `ldap`, but this exclude the `ldap` package from the source code folder

File: LICENSE-binary [hunk 4]
```diff
@@ -322,7 +322,9 @@ org.apache.zookeeper:zookeeper
 
 BSD
 ------------
+org.antlr:antlr-runtime
 org.antlr:antlr4-runtime
+org.antlr:ST4
```
Comment (pan3793) [inline LICENSE-binary:327]: It's required by `kyuubi.authentication.ldap.customLDAPQuery`

File: dev/dependencyList [hunk 5]
```diff
@@ -16,8 +16,10 @@
 #
 
 HikariCP/4.0.3//HikariCP-4.0.3.jar
+ST4/4.3.4//ST4-4.3.4.jar
```
Comment (pan3793) [inline dev/dependencyList:19]: It's required by `kyuubi.authentication.ldap.customLDAPQuery`

**Comment by pan3793** [comment]
CI is green, this is ready for review, cc @wheatxiong

**Comment by pan3793** [comment]
Since it's basically a feature porting from Apache Hive, and all UT&IT passed, I'm going to merge this one in a few days if no oppositions.

**Comment by pan3793** [comment]
Merged to master

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/LdapAuthenticationProviderImpl.scala [hunk 6]
```diff
@@ -41,47 +49,72 @@ class LdapAuthenticationProviderImpl(conf: KyuubiConf) extends PasswdAuthenticat
...
+    }
+
+    var search: DirSearch = null
+    try {
+      search = createDirSearch(bindUser, bindPassword)
+      applyFilter(search, user)
+      if (usedBind) {
+        // If we used the bind user, then we need to authenticate again,
+        // this time using the full user name we got during the bind process.
+        createDirSearch(search.findUserDn(user), password)
```
Comment (dev-lpq) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/LdapAuthenticationProviderImpl.scala:73]: @pan3793, I want to support has domain case. Does the authenticate again process needs to get the username part in the provided user?  If yes, I will create the function `getUserName` and provide a PR.
`  def getUserName(user: String): String =
    if (LdapUtils.hasDomain(user)) LdapUtils.extractUserName(user) else user`

Comment (dev-lpq) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/LdapAuthenticationProviderImpl.scala:73]: ` val username = getUserName(user)
   createDirSearch(search.findUserDn(username), password)`

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/LdapAuthenticationProviderImpl.scala:73]: Sounds make sense to me. The current logic should be same as Apache Hive, would you please check if Hive has same issue? Anyway, go ahead to improve it.

Comment (dev-lpq) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/LdapAuthenticationProviderImpl.scala:73]: sure, Hive branch 4.0.0 has the same issue, I will improve it.

**Pull request closed without merging**